### PR TITLE
Adding loader for Mridangam Stroke Dataset

### PR DESCRIPTION
**Pull request by genisplaja**
### Adding loader for Mridangam Stroke Dataset

 Please use the following title: "Adding loader for MyDATASET". If your pull request is work in progress, change your title to "[WIP] Adding loader for MyDATASET" to avoid reviews while the loader is not ready.

### Description

Please include the following information at the top level docstring for the dataset's module mydataset.py:

- [x] Describe annotations included in the dataset
- [x] Indicate the size of the datasets (e.g. number files and duration, hours)
- [x] Mention the origin of the dataset (e.g. creator, institution)
- [x] Describe the type of music included in the dataset
- [x] Indicate any relevant papers related to the dataset
- [x] Include a description about how the data can be accessed and the license it uses (if applicable)

#### Dataset loaders checklist:

- [x] Create a script in `scripts/`, e.g. `make_my_dataset_index.py`, which generates an index file.
- [x] Run the script on the canonical version of the dataset and save the index in `mirdata/indexes/` e.g. `my_dataset_index.json`.
- [x] Create a module in mirdata, e.g. `mirdata/my_dataset.py`
- [x] Create tests for your loader in `tests/`, e.g. `test_my_dataset.py`
- [x] Add your module to `docs/source/mirdata.rst` and `docs/source/datasets.rst`
- [x] Add the module to `mirdata/__init__.py`
- [x] Add the module to the list in the `README.md` file, section `Currently supported datasets`

This dataset is **fully downloadable** .

#### Please-do-not-edit flag
To reduce friction, we will make commits on top of contributor's pull requests by default unless they use the `please-do-not-edit` flag. If you don't want this to happen don't forget to add the flag when you start your pull request.

### CODE REVIEW

**Comment by nkundiushuti** [comment]
> # [Codecov](https://codecov.io/gh/mir-dataset-loaders/mirdata/pull/285?src=pr&el=h1) Report
> > Merging [#285](https://codecov.io/gh/mir-dataset-loaders/mirdata/pull/285?src=pr&el=desc) into [master](https://codecov.io/gh/mir-dataset-loaders/mirdata/commit/a60d1192f5f2401579ba598c3ea9c7b63ad5db66?el=desc) will **decrease** coverage by `0.07%`.
> > The diff coverage is `96.38%`.
> 
> ```diff
> @@            Coverage Diff             @@
> ##           master     #285      +/-   ##
> ==========================================
> - Coverage   99.01%   98.94%   -0.08%     
> ==========================================
>   Files          22       23       +1     
>   Lines        2340     2455     +115     
> ==========================================
> + Hits         2317     2429     +112     
> - Misses         23       26       +3     
> ```

@genisplaja , I think you are lacking some tests on the methods you implemented for this dataset

**Comment by genisplaja** [comment]
Thanks @nkundiushuti! I fixed the Track class and updated tests for load_stroke_name and load_tonic. Now it should be alright.

**Comment by magdalenafuentes** [comment]
@genisplaja will take a look to this tomorrow. Though there's a small conflict it seems. Btw, thanks for all the work!! :)

File: docs/source/datasets.rst [hunk 1]
```diff
@@ -215,6 +215,19 @@ Vocal Activity
 A time series or sequence of events indicating when singing voice is present in a recording. This type
 of annotation is implicitly available when Vocal F0_ or Vocal Notes_ annotations are available.
 
+.. _stroke-name:
+
+Stroke Name
+^^^^^^^^^^^
+An open "tag" that refers to the original name of the stroke type that is played on a Mridangam instrument.
```
Comment (magdalenafuentes) [inline docs/source/datasets.rst:222]: Maybe we should do this definition general? I can imagine other datasets with stroke types for other instruments, so simply an open tag with the name of the stroke type could be enough

File: docs/source/datasets.rst [hunk 2]
```diff
@@ -215,6 +215,19 @@ Vocal Activity
...
+
+.. _tonic:
+
+Tonic
+^^^^^^^^^^^
+Stored as metadata, the absolute tonic of a track. It is a useful metre for tracks with a single stroke or
+note. Originally added to describe the tonic of the Mridangam strokes.
```
Comment (magdalenafuentes) [inline docs/source/datasets.rst:229]: Same comment here about being general

File: mirdata/jams_utils.py [hunk 3]
```diff
@@ -86,7 +87,12 @@ def jams_converter(
     # metadata
     if metadata is not None:
         for key in metadata:
-            if key == 'duration' and duration is not None and metadata[key] != duration:
+            if (
+                key == 'duration'
+                and duration is not None
+                and metadata[key] != duration
+                and 'spectrum' not in audio_path
```
Comment (magdalenafuentes) [inline mirdata/jams_utils.py:94]: What is `spectrum` here?

File: mirdata/jams_utils.py [hunk 4]
```diff
@@ -554,3 +572,28 @@ def tag_gtzan_to_jams(tags):
...
+    """
+    jannot_tag_open = jams.Annotation(namespace='tag_open')
+    jannot_tag_open.annotation_metadata = jams.AnnotationMetadata(data_source='mirdata')
+    if tags[0] is not None:
+        if not isinstance(tags[0], str):
+            raise TypeError('Type should be str.')
+        jannot_tag_open.append(time=0.0, duration=0.0, value=tags[0])
+    if tags[1] is not None:
+        jannot_tag_open.sandbox = jams.Sandbox(name=tags[1])
+    return jannot_tag_open
```
Comment (magdalenafuentes) [inline mirdata/jams_utils.py:599]: Nice! Thanks for this!

File: mirdata/mridangam_stroke.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+    'remote_data': download_utils.RemoteFileMetadata(
+        filename='mridangam_stroke_1.5.zip',
+        url='https://zenodo.org/record/4068196/files/mridangam_stroke_1.5.zip?download=1',
+        checksum='39af55b2476b94c7946bec24331ec01a',  # the md5 checksum
+        destination_dir=None,  # relative path for where to unzip the data, or None
+    ),
+}
+
+DATA = utils.LargeData(
+    'mridangam_stroke_index.json'
```
Comment (magdalenafuentes) [inline mirdata/mridangam_stroke.py:60]: I assume there isn't metadata right?

File: mirdata/mridangam_stroke.py [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+            data_home = utils.get_default_dataset_path(DATASET_DIR)
+
+        self._data_home = data_home
+        self._track_paths = DATA.index[track_id]
+
+        self.audio_path = os.path.join(self._data_home, self._track_paths['audio'][0])
+
+    @utils.cached_property
+    def stroke_name(self):
+        """(String): audio signal, sample rate"""
```
Comment (magdalenafuentes) [inline mirdata/mridangam_stroke.py:90]: update :)

File: mirdata/mridangam_stroke.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+        return load_stroke_name(self.audio_path)
+
+    @utils.cached_property
+    def tonic(self):
+        """(np.ndarray, float): audio signal, sample rate"""
```
Comment (magdalenafuentes) [inline mirdata/mridangam_stroke.py:95]: same with this one

File: scripts/make_giantsteps_key.py [hunk 8]
```diff
@@ -57,5 +57,7 @@ def main(args):
 
 if __name__ == '__main__':
     PARSER = argparse.ArgumentParser(description='Make giantsteps_key index file.')
-    PARSER.add_argument('giantsteps_key_data_path', type=str, help='Path to giantsteps_key data folder.')
+    PARSER.add_argument(
+        'giantsteps_key_data_path', type=str, help='Path to giantsteps_key data folder.'
+    )
```
Comment (magdalenafuentes) [inline scripts/make_giantsteps_key.py:62]: This is being edited by black though normally it would not be part of the PR. Maybe try in future PRs to run `black -S` and will skip strings. We will fix this in the Docs audit later on #245

**Review by magdalenafuentes — DISMISSED** [review]
Looking really great @genisplaja! Just left a couple nit comments and it's ready to go :)

File: mirdata/jams_utils.py (hunk 3 above)
Comment (nkundiushuti) [reply mirdata/jams_utils.py:94]: @genisplaja is this a mixup from cofla cante100 dataset? there is no spectrum in this dataset

File: docs/source/datasets.rst (hunk 2 above)
Comment (nkundiushuti) [reply docs/source/datasets.rst:229]: @magdalenafuentes should we define an ontology and use it everywhere?

File: scripts/print_track_docstring.py [hunk 9]
```diff
@@ -46,9 +46,9 @@ def get_attributes_and_properties(class_instance):
         else:
             raise ValueError("Unknown type {}".format(attr))
 
-    non_attributes = list(itertools.chain.from_iterable(
-        [properties, cached_properties, functions]
-    ))
+    non_attributes = list(
+        itertools.chain.from_iterable([properties, cached_properties, functions])
```
Comment (nkundiushuti) [inline scripts/print_track_docstring.py:50]: maybe it's a good idea to remove from PR files which are not related to the dataset, even if it's just formatting. it makes it more cumbersome to check for errors on large files.

File: docs/source/datasets.rst (hunk 2 above)
Comment (magdalenafuentes) [reply docs/source/datasets.rst:229]: Yes! I also added a comment on #245 to explain this better in the docs later

File: mirdata/mridangam_stroke.py (hunk 5 above)
Comment (genisplaja) [reply mirdata/mridangam_stroke.py:60]: No! There is just a file for each folder to relate the files with their respective Freesound link (because all the files are uploaded to Freesound).

File: mirdata/mridangam_stroke.py [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+        audio_path (str): Local path where the track is stored.
+    Returns:
+        stroke_name (str): stroke name or type extracted from filename
+    """
+    if not os.path.exists(audio_path):
+        raise IOError("audio_path {} does not exist".format(audio_path))
+
+    if '.wav' in audio_path:
+        audio_data = audio_path.split('__')[2]
+        stroke_name = audio_data.split('-')[0]
```
Comment (nkundiushuti) [inline mirdata/mridangam_stroke.py:205]: I have doubts with this
an example of a file name is: 225486__akshaylaya__dheem-c-078.wav 
I think the stroke name here is dheem while the tonic is C. is that correct?
I think it would be better if we validate the stroke names against a pre-defined dictionary of strokes:
`assert stroke_name is in self.stroke_dictionary, "Stroke {} not in stroke dictionary".format(stroke_name)`
where self.stroke_dictionary can be defined in __init__ comprising all the possible strokes

File: mirdata/mridangam_stroke.py [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+        audio_path (str): Local path where the track is stored.
+    Returns:
+        tonic (str): track tonic extracted from filename
+    """
+    if not os.path.exists(audio_path):
+        raise IOError("audio_path {} does not exist".format(audio_path))
+
+    if '.wav' in audio_path:
+        audio_data = audio_path.split('__')[2]
+        tonic_info = audio_data.split('-')[1]
```
Comment (nkundiushuti) [inline mirdata/mridangam_stroke.py:222]: similar to stroke, it would be useful to validate against an existing tonic dictionary 
`assert tonic_info is in self.tonic_dictionary `

File: scripts/make_giantsteps_key.py (hunk 8 above)
Comment (nkundiushuti) [reply scripts/make_giantsteps_key.py:62]: you can do this to exclude it, provided that origin/master is up to date
```
git checkout origin/master -- scripts/giantsteps_key.py
git commit
git push
```

**Review by nkundiushuti — CHANGES_REQUESTED** [review]
some useful changes: assert tonic and stroke parsed from file name against a pre-defined dictionary (defined in __init__)

File: mirdata/mridangam_stroke.py (hunk 11 above)
Comment (genisplaja) [reply mirdata/mridangam_stroke.py]: Done for both!! Thanks for the review!

**Review by nkundiushuti — APPROVED** [review]

**Review by rabitt — COMMENTED** [review]
.

**Comment by rabitt** [comment]
@genisplaja I'm reviewing now and found some small bugs! will post my review in the next few minutes!

File: mirdata/mridangam_stroke.py [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+    if data_home is None:
+        data_home = utils.get_default_dataset_path(DATASET_DIR)
+
+    data = {}
+    for key in DATA.index.keys():
+        data[key] = Track(key, data_home=data_home)
+    return data
+
+
+def load_stroke_name(audio_path):
```
Comment (rabitt) [inline mirdata/mridangam_stroke.py:205]: In general, our convention is that `load_*` methods should be loading content from a file, rather than parsing strings. Can you make these both Track attributes defined in `__init__`? `self.stroke_name = ...` and `self.tonic = ...`

Also, since the filepath can be parsed even if the audio file doesn't exist,  we don't need to check if it exists.

File: mirdata/mridangam_stroke.py [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+
+        if data_home is None:
+            data_home = utils.get_default_dataset_path(DATASET_DIR)
+
+        self._data_home = data_home
+        self._track_paths = DATA.index[track_id]
+
+        self.audio_path = os.path.join(self._data_home, self._track_paths['audio'][0])
+
+    @utils.cached_property
```
Comment (rabitt) [inline mirdata/mridangam_stroke.py:98]: see the below comment, but both of these cached properties can just be defined in the __init__ and can be removed here :)

**Review by rabitt — CHANGES_REQUESTED** [review]
New loader checklist:
- [x] Create a script in `scripts/`, e.g. `make_my_dataset_index.py`
- [x] Run the script on the canonical version of the dataset and save the index in `mirdata/indexes/` e.g. `my_dataset_index.json`
- [x] Create a module in `mirdata`, e.g. `mirdata/my_dataset.py`
- [x] Create tests in `tests/`, e.g. `test_my_dataset.py`
- [x] Add the module to `docs/source/mirdata.rst`
- [x] Add the module to `mirdata/__init__.py`
- [x] Add the module to the list in the README.md file, section Currently supported datasets
- [x] Run test_full_dataset.py
- [x] Make sure the docs build properly

Two small comments (the docs, and simplifying the Track attributes) and then this is ready!!

File: mirdata/mridangam_stroke.py (hunk 12 above)
Comment (genisplaja) [reply mirdata/mridangam_stroke.py]: Fixed!

File: mirdata/mridangam_stroke.py [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+
+
+class Track(track.Track):
+    """Mridangam Stroke track class
+    Args:
+        track_id (str): track id of the track
+        data_home (str): Local path where the dataset is stored.
+            If `None`, looks for the data in the default directory, `~/mir_datasets/Mridangam-Stroke`
+    Attributes:
+        track_id (str): track id
```
[the remark this answers is not in the document]
Comment (genisplaja) [reply mirdata/mridangam_stroke.py:81]: Added! I also added the stroke name and tonic!

File: docs/source/datasets.rst [hunk 15]
```diff
@@ -73,6 +71,9 @@ click the link provided in the "Module" column.
 | medleydb_pitch_   | | MedleyDB Pitch    | - audio: 🔑         | - :ref:`f0`               | 103    |
 |                   | | Tracking Subset   | - annotations: ✅   | - :ref:`instruments`      |        |
 +-------------------+---------------------+---------------------+---------------------------+--------+
+| mridangam_stroke_ | Mridangam Stroke    | - audio: ✅         | - :ref:`stroke_name`      | 6977   |
+|                   |                     | - annotations: ✅   | - :ref:`tonic`            |        |
++-------------------+------------------- -+---------------------+---------------------------+--------+
```
Comment (rabitt) [inline docs/source/datasets.rst:76]: still not building :( there's a ` ` instead of a `-` near the end of column 2

(you can check this by clicking on the `Details` in the readthedocs test)

**Review by rabitt — APPROVED** [review]
The table in the docs is still breaking but it's a one character fix and then let's merge!

File: docs/source/datasets.rst (hunk 15 above)
Comment (genisplaja) [reply docs/source/datasets.rst]: That's right, found it!! As my table did not throw any error I didn't see it. Fixed now!

**Comment by rabitt** [comment]
The docs still aren't working but I dont see why! I'm merging this and I'll fix in a separate PR.

Thanks @genisplaja !
